src/coreComponents/linearAlgebra/unitTests/testLinearSolverParametersEnums.cpp

```diff
@@ -0,0 +1,67 @@
+/*
+ * ------------------------------------------------------------------------------------------------------------
+ * SPDX-License-Identifier: LGPL-2.1-only
+ *
+ * Copyright (c) 2018-2020 Lawrence Livermore National Security LLC
+ * Copyright (c) 2018-2020 The Board of Trustees of the Leland Stanford Junior University
+ * Copyright (c) 2018-2020 Total, S.A
+ * Copyright (c) 2020-     GEOSX Contributors
+ * All right reserved
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+#include "linearAlgebra/utilities/LinearSolverParameters.hpp"
+
+#include "linearAlgebra/unitTests/testLinearAlgebraUtils.hpp"
+
+#include <gtest/gtest.h>
+
+using namespace geosx;
+
+// The `ENUM_STRING` implementation relies on consistency between the order of the `enum`,
+// and the order of the `string` array provided. Since this consistency is not enforced, it can be corrupted anytime.
+// This unit test aims at preventing from this implicit relationship to bring a bug.
+
+TEST( LinearSolverParametersEnums, SolverType )
+{
+  using EnumType = LinearSolverParameters::SolverType;
+
+  ASSERT_EQ( "direct", toString( EnumType::direct ) );
+  ASSERT_EQ( "cg", toString( EnumType::cg ) );
+  ASSERT_EQ( "gmres", toString( EnumType::gmres ) );
+  ASSERT_EQ( "fgmres", toString( EnumType::fgmres ) );
+  ASSERT_EQ( "bicgstab", toString( EnumType::bicgstab ) );
+  ASSERT_EQ( "preconditioner", toString( EnumType::preconditioner ) );
```

Comment:
> although we just went from single to double definition of the strings, effectively.

Actually we currently have an `enum -> int` mapping and a `int -> string` mapping that we "compose".
I've decided to to validate the final composed `enum -> string` mapping.
I hesitated to validate the `enum -> int` because I thought it may have been less clear, and because the final goal is to have `enum -> string`. I'm not sure.

And it's not really double definition, there is one definition and one expected result.
_And by the design of the test, any discrepancy can't go unnoticed (which is the critical point for me)._

> Overall, I'm still looking for a better implementation of the feature

Using http://jhnet.co.uk/articles/cpp_magic (see raw code [here](http://jhnet.co.uk/articles/cpp_magic/map_working.txt)) and https://renenyffenegger.ch/notes/development/languages/C-C-plus-plus/preprocessor/macros/__VA_ARGS__/count-arguments (because the `HAS_ARGS` impl of first example did not work in my precise case, I did not have time to investigate more), then

```C++
#define _MY_CASE_LINE(k, v) case k: return v;

#define MY_CASE_LINE(p) _MY_CASE_LINE p

EVAL( MAP( MY_CASE_LINE, ( 0, "a" ), ( 1, "b" ), ( 2, "c" ), ( 3, "d" ) ) )
```
expands into (line breaks are mine)
```c++
case 0: return "a";
case 1: return "b";
case 2: return "c";
case 3: return "d";
```
which eventually will let you compile with `-Wswitch-enum` (see [gcc's doc](https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html)).
This will help you detect any new `enum` not considered. (You may even play with `#pragma ??? diagnostic...`).